[PULL REQUEST] Convert input.geos to YAML format; save rundir_vars.txt in the runConfig subdirectory - Closes #1111

yantosca opened the pull request:
This is a draft PR corresponding to the discussion in #1111.

We are in the process of porting GEOS-Chem's main configuration file (`input.geos`) to YAML format.  The working file name that I am using is `input_options.yml` but this can be changed if need be.

Also bundled with this PR will be the moving of the `rundir_vars.txt` file (which can be used for automatic rundir generation) to `rundirConfig/rundir_vars.txt`.  This file will be saved in the rundir Git repo as well.

Lastly, we will remove the `rundir.version` file and instead append that as a header to `rundir_vars.txt`.  The new `rundir_vars.txt` will look similar to this:

```
#  This run directory was created with:
#  /local/ryantosca/GC/gcc_dev/src/GEOS-Chem/run/GCClassic/createRunDir.sh.
# 
#  GEOS-Chem repository version information:
# 
#   Remote URL: https://github.com/geoschem/geos-chem.git
#   Branch: feature/yaml-input-options
#   Commit: Default GCClassic rundir names now include extra grid information
#   Date: Wed Feb 2 10:39:40 2022 -0500
#   User: Bob Yantosca
#   Hash: 16c7fd59a
# 
# 
# Changes to the following run directory files are tracked by git:
# 
#  [master (root-commit) 0110a4f] Initial run directory
#  13 files changed, 12378 insertions(+)
#  create mode 100644 HEMCO_Config.rc
#  create mode 100644 HEMCO_Diagn.rc
#  create mode 100644 HISTORY.rc
#  create mode 100755 archiveRun.sh
#  create mode 100755 cleanRunDir.sh
#  create mode 100755 download_data.py
#  create mode 100644 download_data.yml
#  create mode 100755 geoschem.benchmark.run
#  create mode 100755 getRunInfo
#  create mode 100644 input_options.yml
#  create mode 100755 metrics.py
#  create mode 100644 rundirConfig/rundir_vars.txt
#  create mode 100644 species_database.yml
# 
# 
#  IMPORTANT: ONLY USE THESE RUNDIR SETTINGS WITH THIS COMMIT!






RUNDIR_ACID_UPTAKE='false'
RUNDIR_AERO_BCPI='* AERO_BCPI      $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_BCPI  2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_AERO_BCPO='* AERO_BCPO      $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_BCPO  2010-2019/1-12/1/0 C xyz 1        * - 1 1'
RUNDIR_AERO_DST1='* AERO_DST1      $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.SpeciesConc.$YYYY$MM$DD_0000z.nc4      SpeciesConc_DST1  2010-2019/1-12/1/0 C xyz 1        * - 1 1'
... etc ...
```

lizziel commented:
Thanks @yantosca! I would prefer a name that includes geoschem so that when it is included in other models it is clear what module the config file goes with. How does everyone feel about simply `geoschem.yml`?

yantosca commented:
Thanks @lizziel.  I recall that that @christophkeller suggested to avoid using geoschem because that gets confusing with the gridded component name (such as GEOSCHEMchem).  I am open to `geoschem.yml` or `geoschem_config.yml` though, as that is in line with `HEMCO_Config`.

yantosca commented:
@msulprizio @lizziel @christophkeller here is the original comment
https://github.com/geoschem/geos-chem/discussions/1111#discussioncomment-2091974

laestrada commented:
Hi @yantosca, looks great so far! 

1. I think that I second the suggestion of @msulprizio `geoschem_config.yml` or `geoschem_options.yml`. Maybe we could also add a comment in the file that the configuration is used at runtime and does not require recompilation?

2. The categories look good to me, but I wonder if the var `met_field` should be renamed `met_dir`? To be consistent with. `root_data_dir`? Unless I am misinterpreting the purpose of `met_field`

yantosca commented:
@laestrada, thanks.  It was "met field" in the old input.geos file so I ported that over.  I can change it to met_dir if you think that's more descriptive.

Also I can switch to `geoschem_config.yml`, I think that is probably the most clear name we can use.

yantosca commented:
@laestrada... also, if you find any confusing tag names in the YAML file let me know.  I tried to make them line up as much as possible with the tags in the input.geos file.  I can also add comments later on as well.

laestrada commented:
@yantosca Sounds good -- thanks, I'll take a closer look at the tag names! I do think `met_dir` is more descriptive, especially since the corresponding symlink in the rundir is `MetDir`. However, if anyone feels strongly towards the original name, I can see the logic in trying to avoid confusion by maintaining the same names

msulprizio commented:
I believe that should stay `met_field` because it's used to define the meteorology field used throughout GEOS-Chem. It's not actually used for the MetDir -- that's defined in HEMCO_Config.rc.

yantosca commented:
> I believe that should stay `met_field` because it's used to define the meteorology field used throughout GEOS-Chem. It's not actually used for the MetDir -- that's defined in HEMCO_Config.rc.

Thanks @msulprizio, I kind of thought that.

laestrada commented on run/GCClassic/input_options.yml.templates/input_options.yml.fullchem:
```diff
@@ -0,0 +1,426 @@
+    allow_homogeneous_NAT: false
+    NAT_supercooling_req_in_K: 3.0
+    supersat_factor_req_for_ice_nucl: 1.2
+    calc_strat_aero_optdepth: true
```
maybe calc_strat_aod for the sake of consistency with `aod_wavelengths_in_nm`?

yantosca replied:
Now fixed.

yantosca replied:
This is now fixed in https://github.com/geoschem/geos-chem/commit/a6cc6617a6ecb9597e8d39ce7ceff0eb880739e4

laestrada commented on run/GCClassic/input_options.yml.templates/input_options.yml.fullchem:
```diff
@@ -0,0 +1,426 @@
+timesteps:
+  transport_timestep_in_s: ${RUNDIR_TRANSPORT_TS}
+  chemistry_timestep_in_s: ${RUNDIR_CHEMISTRY_TS}
+  radiation_timestep_in_s: 10800
```
perhaps `secs` instead of `in_s`? eg. `radiation_timestep_secs`

laestrada commented on run/GCClassic/input_options.yml.templates/input_options.yml.fullchem:
```diff
@@ -0,0 +1,426 @@
+     use_non_local_pbl: ${RUNDIR_USE_NLPBL}
+
+  photolysis:
+    input_directory: ${RUNDIR_DATA_ROOT}/CHEM_INPUTS/FAST_JX/v2021-10/
```
Perhaps `input_dir` for consistency with `root_data_dir` above?

yantosca replied:
@laestrada, good catch.  We can make it input_dir.

yantosca replied:
Now fixed.

yantosca replied:
Also fixed in a6cc6617a6ecb9597e8d39ce7ceff0eb880739e4

laestrada commented on run/GCClassic/input_options.yml.templates/input_options.yml.fullchem:
```diff
@@ -0,0 +1,426 @@
+  transport:
+    activate: true
+    fill_negative_values: true
+    iord_jord_kord: [3, 3, 7]
```
Again maybe just ignorance on my part, but I am not sure what the `ord` refers to here

yantosca replied:
@laestrada, those are options that control the transport in GEOS-Chem.  These are listed in the code: 

https://github.com/geoschem/geos-chem/blob/d8701729063383eb7c38d0f18619f0a459aac0b7/GeosCore/tpcore_fvdas_mod.F90#L26-L67

Maybe we need to make these clear with some comments in the yaml file.

yantosca replied:
actually, only GEOS-Chem "Classic".  Not needed for FV3.

yantosca replied:
Changed tags to `tpcore_fill_negative_values` and `tpcore_iord_jord_kord` to denote that they are for TPCORE only (GEOS-Chem Classic)

yantosca replied:
Fixed in a6cc6617a6ecb9597e8d39ce7ceff0eb880739e4

christophkeller commented:
`geoschem_config.yml` is fine with me!

yantosca commented on run/GCClassic/input_options.yml.templates/input_options.yml.fullchem:
```diff
@@ -0,0 +1,426 @@
+      percent_channel_A_HONO: 66.667
+      percent_channel_B_NO2: 33.333
+
+  rrtmg:
```
@laestrada, RRTMG is the radiative transfer model that a lot of the aerosols people use.  It can provide radiative forcing fluxes as output (although it adds a lot of overhead).  I can make that "rrtmg_rad_forcing" or something clearer if you wish.

yantosca replied:
Changed the tag to `rrtmg_rad_transfer_model`

yantosca replied:
Also fixed in a6cc6617a6ecb9597e8d39ce7ceff0eb880739e4

lizziel commented on run/GCClassic/input_options.yml.templates/input_options.yml.fullchem:
```diff
@@ -0,0 +1,426 @@
+#      JMIN_and_JMAX_of_region: [1, 46]
+#      LMIN_and_LMAX_of_region: [1, 1]
+#
+#    ND51b_satellite:
```
Could this be more descriptive of how ND51b differs from ND51?

yantosca replied:
We can.  But if we get the PR #1134 into 14.0.0, that will make these obsolete.

lizziel commented on run/GCClassic/input_options.yml.templates/input_options.yml.fullchem:
```diff
@@ -0,0 +1,426 @@
+    all_sky_flux: ${RUNDIR_RRTMG_OPTS}
+
+  transport:
+    activate: true
```
This section is tough since it lists advected species but activate will be false for GCHP. That makes it seem the details in the section are not applicable in GCHP. Would it be possible to omit GC-Classic-only lines in the GCHP yml, such as transport activate? Alternatively, maybe we can figure out how best to indicate which lines are GC-Classic only. And yet another option is move the advection species to their own section and label this one gcclassic_transport.

yantosca replied:
We should be able to omit fields that do not apply to GCHP -- we don't have to include them in the GCHP rundir geoschem_config.yml file. Fields that are not found are set to missing values (e.g. false, or -999, etc.)

Also I think we could do something like this:

```yaml
  transport:
    gcclassic_tpcore:                 # GEOS-Chem Classic only
      activate: true                  # GEOS-Chem Classic only
      fill_negative_values: true      # GEOS-Chem Classic only
      iord_jord_kord: [3, 3, 7]       # GEOS-Chem Classic only
    transported_species:
      - ACET
      - ACTA
      ..etc...
    passive_species:
      PassiveTracer:
        long_name: Passive_tracer_for_mass_conservation_evaluation
        mol_wt_in_g: 1.0
        lifetime_in_s: -1
        default_bkg_conc_in_vv: 1.0e-7
      SF6Tracer:
        long_name: Sulfur_hexafluoride
        mol_wt_in_g: 146.0
        lifetime_in_s: -1
        default_bkg_conc_in_vv: 1.0e-20
      CH3ITracer:
        long_name: Methyl_iodide
        mol_wt_in_g: 142.0
        lifetime_in_s: 4.32e5
        default_bkg_conc_in_vv: 1.0e-20
```

yantosca replied:
this is just an example to show how you can add passive species to any simulation if you want.

yantosca replied:
Now fixed in a6cc6617a6ecb9597e8d39ce7ceff0eb880739e4 and 973c2266ccf64e8d587d1fa705eef60ba2896056

yantosca commented on run/GCClassic/input_options.yml.templates/input_options.yml.fullchem:
```diff
@@ -0,0 +1,426 @@
+#============================================================================
+# Simulation settings
+#============================================================================
+simulation:
```
@lizziel, I believe that we can!

yantosca commented:
The geoschem_config.yml for the GCClassic TransportTraces simulation is now complete.  We now only need to include the switches that we need, see: https://github.com/geoschem/geos-chem/blob/feature/yaml-input-options/run/GCClassic/geoschem_config.yml.templates/geoschem_config.yml.TransportTracers

yantosca commented:
The latest updates to the `feature/yaml-input-options` pass GEOS-Chem Classic integration tests:

```console
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 48
Execution tests failed: 0
Execution tests not yet completed: 0
```
GCHP integration tests pending

yantosca commented:
GCHP integration tests failed in `Interfaces/GCHP/Chem_GridCompMod.F90` because it was expecting to get the advected species names from `input.geos` instead of `geoschem_config.yml`.  I will work on this next.

yantosca commented:
I updated this branch so that GCHP simulations now read from `geoschem_config.yml` files.  These updates pass all GCHP integration tests.
```console
==============================================================================
GCHP: Execution Test Results

Number of execution tests: 3
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gchp_fullchem_benchmark_merra2_c48...............Execute Simulation.....PASS
gchp_fullchem_standard_merra2_c24................Execute Simulation.....PASS
gchp_TransportTracers_geosfp_c24.................Execute Simulation.....PASS
 
Summary of execution test results:
------------------------------------------------------------------------------
Execution tests passed:        3
Execution tests failed:        0
Execution tests not completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

yantosca commented:
All GC-Classic integration tests also passed.
```console
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 48
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

yantosca commented:
Also note, this PR should go in after #990 (the update for 4-D to 3-D species arrays).

yantosca commented:
GCHP simulations for fullchem-benchmark and transport tracers give identical results w/r/t equivalent simulations using input.geos:

**TransportTracers**
```console
(base) [holyjacob01 tests]$ ./compare_diags.py gchptt.yml 
Using configuration file gchptt.yml
... Printing totals and differences
Variable               Ref=input.geos           Dev=YAML config          Dev - Ref
ARCHV_DRY_TOTN       : 0.0                    | 0.0                    | 0.0 
ARCHV_WET_TOTN       : 0.0                    | 0.0                    | 0.0 
AREA                 : 510064471965696.0      | 510064471965696.0      | 0.0 
DEP_RESERVOIR        : 0.0                    | 0.0                    | 0.0 
DRYPERIOD            : 0.0                    | 0.0                    | 0.0 
DryDepNitrogen       : 0.0                    | 0.0                    | 0.0 
GCCTROPP             : 587900.9               | 587900.9               | 0.0 
GWET_PREV            : 0.0                    | 0.0                    | 0.0 
LAI_PREVDAY          : 0.0                    | 0.0                    | 0.0 
PARDF_DAVG           : 0.0                    | 0.0                    | 0.0 
PARDR_DAVG           : 0.0                    | 0.0                    | 0.0 
PFACTOR              : 0.0                    | 0.0                    | 0.0 
T_DAVG               : 0.0                    | 0.0                    | 0.0 
T_PREVDAY            : 0.0                    | 0.0                    | 0.0 
TropLev              : 112215.0               | 112215.0               | 0.0 
WetDepNitrogen       : 0.0                    | 0.0                    | 0.0 
BXHEIGHT             : 268750256.6461804      | 268750256.6461804      | 0.0 
DELP_DRY             : 3402627.2097128914     | 3402627.2097128914     | 0.0 
H2O2AfterChem        : 0.0                    | 0.0                    | 0.0 
IsorropBisulfate     : 0.0                    | 0.0                    | 0.0 
IsorropHplusFine     : 0.0                    | 0.0                    | 0.0 
IsorropNitrateFine   : 0.0                    | 0.0                    | 0.0 
IsorropSulfate       : 0.0                    | 0.0                    | 0.0 
KPPHvalue            : 0.0                    | 0.0                    | 0.0 
SO2AfterChem         : 0.0                    | 0.0                    | 0.0 
SPC_Be10             : 2.1828822934716557e-13 | 2.1828822934716557e-13 | 0.0 
SPC_Be10Strat        : 2.140991574997023e-13  | 2.140991574997023e-13  | 0.0 
SPC_Be7              : 1.2566909447168238e-14 | 1.2566909447168238e-14 | 0.0 
SPC_Be7Strat         : 1.2213985825023085e-14 | 1.2213985825023085e-14 | 0.0 
SPC_CH3ITracer       : 2.637966254822513e-15  | 2.637966254822513e-15  | 0.0 
SPC_CLOCK            : 487775437173465.2      | 487775437173465.2      | 0.0 
SPC_COAnthroEmis25dayTracer : 0.0012361103232730868  | 0.0012361103232730868  | 0.0 
SPC_COAnthroEmis50dayTracer : 0.0022502489432821666  | 0.0022502489432821666  | 0.0 
SPC_COUniformEmis25dayTracer : 5.12296583738645       | 5.12296583738645       | 0.0 
SPC_PassiveTracer    : 0.024882990775128334   | 0.024882990775128334   | 0.0 
SPC_Pb210            : 1.593802476399884e-15  | 1.593802476399884e-15  | 0.0 
SPC_Pb210Strat       : 6.143205559697142e-17  | 6.143205559697142e-17  | 0.0 
SPC_Rn222            : 7.994371948751801e-16  | 7.994371948751801e-16  | 0.0 
SPC_SF6Tracer        : 1.77989039169885e-06   | 1.77989039169885e-06   | 0.0 
SPC_default_bkg_conc_in_vv: 1.0e : nan                    | 0.0                    | nan 
SPC_lifetime_in_s:   : nan                    | 0.0                    | nan 
STATE_PSC            : 0.0                    | 0.0                    | 0.0 
```

**fullchem-benchmark** (NOTE: only showing nonzero diffs due to large # of species)
```console
Using configuration file gchpbmk.yml
... Printing totals and differences
... Only showing variables with |absolute difference| > 0.0
Variable               Ref=input.geos           Dev=YAML config          Dev - Ref
```

yantosca commented:
NOTE: Also in this PR, we remove the bpch diagnostics for the Hg simulatons, as they are no longer needed.

msulprizio commented:
@lizziel @laestrada I noticed your reviews were still pending. Are your concerns mentioned in your initial reviews above now addressed?

yantosca commented:
> @yantosca While merging with the latest dev branch, I noticed several remaining uses of `input.geos` that I fixed in the above commits.
> 
> I also noted that `download_data.py` and `getRunInfo` (used by GCClassic's `createRunDir.sh`) still referenced fields in input.geos and updated those. Could you review those changes and make sure those scripts still work?
> 
> I am running integration tests on these fixes now.

Thanks @msulprizio for the fixes. 

I'll test the scripts right away.  Stay tuned.

laestrada commented:
I believe my comments were addressed as well

yantosca commented:
@msulprizio, also note there are modifications to the CMakeLists in the feature/yaml-input-options branch in the superprojects.  This doesn't contain any submodule hashes so it should be OK to merge in.  (Basically have to look for geoschem_config.yml instead of input.geos to make sure it's a valid run dir.)

yantosca commented:
@msulprizio, I tested the download_data.py script on the AWS cloud and it pulled over all the data needed.   Good to go.

msulprizio commented:
All GCClassic and GCHP integration tests on this branch passed:

```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

Using 24 OpenMP threads
Number of execution tests: 48
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gc_025x03125_CH4_geosfp_47L_na...................Execute Simulation.....PASS
...etc...

Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 48
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

```console
==============================================================================
GCHP: Execution Test Results

Number of execution tests: 3
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gchp_fullchem_benchmark_merra2_c48...............Execute Simulation.....PASS
gchp_fullchem_standard_merra2_c24................Execute Simulation.....PASS
gchp_TransportTracers_geosfp_c24.................Execute Simulation.....PASS
 
Summary of execution test results:
------------------------------------------------------------------------------
Execution tests passed:        3
Execution tests failed:        0
Execution tests not completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```
